### PR TITLE
Use Formik for login form

### PR DESCRIPTION
**Pull request by andemerie**

### CODE REVIEW

File: components/auth/FormField.js [hunk 1]
```diff
@@ -5,12 +5,12 @@ import Icon from 'components/common/Icon';
 import Text from 'components/common/Text';
 
 /* eslint-disable jsx-a11y/label-has-for */
```
Comment (Drapegnik) [inline components/auth/FormField.js]: please, try to remove this line

File: components/auth/LoginForm.js [hunk 2]
```diff
@@ -1,29 +1,43 @@
...
+import Form from 'components/common/Form';
+import Button from 'components/common/Button';
 import Text from 'components/common/Text';
 
+import { actions } from 'redux/ducks/auth';
+
 import FormField from './FormField';
 
-const fields = {
+const loginInitialValues = {
```
Comment (Drapegnik) [inline components/auth/LoginForm.js]: how about `INITIAL_FORM` name?

File: components/auth/FormField.js [hunk 3]
```diff
@@ -5,12 +5,12 @@ import Icon from 'components/common/Icon';
...
-const FormField = ({ inputId, label, icon, children, pending, touched, error, successMessage }) => {
+const FormField = ({ id, label, icon, children, pending, touched, error, successText }) => {
```
Comment (Drapegnik) [inline components/auth/FormField.js:8]: lets extract it to `common` components & add to storybook

File: components/auth/LoginForm.js [hunk 4]
```diff
@@ -35,137 +49,131 @@ const fields = {
...
-  );
-};
+              );
+            })}
+          <div className="field">
+            <div className="control has-text-centered">
+              <Button
+                className="button is-uppercase login__button"
+                type="submit"
+                disabled={hasErrors(errors, touched) || isSubmitting}
```
Comment (Drapegnik) [inline components/auth/LoginForm.js]: please, move this logic to `<Button />`

File: components/auth/LoginForm.js [hunk 5]
```diff
@@ -35,137 +49,131 @@ const fields = {
...
+              </Button>
+            </div>
+          </div>
+        </form>
+      )}
+    </Form>
+  </div>
+);
+
+const errorsPropsTypes = {};
```
Comment (Drapegnik) [inline components/auth/LoginForm.js]: remove this

File: components/auth/LoginForm.js [hunk 6]
```diff
@@ -35,137 +49,131 @@ const fields = {
...
-    successMessage: 'auth.passwordsEqual',
+    successText: 'auth.passwordsEqual',
+    type: 'password',
   },
 };
 
-const errorValidator = values => {
+const loginFieldsKeys = Object.keys(loginFields);
+
+const loginValidator = values => {
```
Comment (Drapegnik) [inline components/auth/LoginForm.js:68]: please pass validator direct to `Field`

File: components/auth/LoginForm.js [hunk 7]
```diff
@@ -35,137 +49,131 @@ const fields = {
...
+const LoginForm = ({ url, allowSignUp, pending }) => (
+  <div>
+    <h1 className="title is-size-5 has-text-centered">
+      <Text id="auth.signIn" />
+    </h1>
+    <Form
+      initialValues={loginInitialValues}
+      validate={loginValidator}
+      action={actions.signIn}
+      handleSubmit={(dispatchAction, { signUp, ...userData }, { setSubmitting }) => {
```
Comment (Drapegnik) [inline components/auth/LoginForm.js]: redundant, `Form` should receive only `action` to submit

File: components/auth/LoginForm.js [hunk 8]
```diff
@@ -35,137 +49,131 @@ const fields = {
...
-                  </div>
-                </div>
-              )}
-              {!allowSignUp && (
-                <div className="signup-not-available">
-                  <Text id="auth.signup-not-available" />
-                </div>
-              )}
-              {keys.filter(key => formApi.values.signUp || !fields[key].onlyOnSignUp).map(key => (
+const LoginForm = ({ url, allowSignUp, pending }) => (
```
Comment (Drapegnik) [inline components/auth/LoginForm.js]: please omit `pending`, remove it from reducer, and use `Formik` internal state `isSubmitting` for handling this logic

File: components/auth/LoginForm.js [hunk 9]
```diff
@@ -35,137 +49,131 @@ const fields = {
...
+    <h1 className="title is-size-5 has-text-centered">
+      <Text id="auth.signIn" />
+    </h1>
+    <Form
+      initialValues={loginInitialValues}
+      validate={loginValidator}
+      action={actions.signIn}
+      handleSubmit={(dispatchAction, { signUp, ...userData }, { setSubmitting }) => {
+        const {
+          query: { lang, next = `/${lang}/articles` },
```
Comment (Drapegnik) [inline components/auth/LoginForm.js]: you can pass `next` instead of `url` to `LoginForm`

File: components/common/Form.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+import { Formik } from 'formik';
+
+// eslint-disable-next-line react/prefer-stateless-function
+class Form extends Component {
```
Comment (Drapegnik) [inline components/common/Form.js]: please change to functional component

File: components/common/Form.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  };
+
+  render() {
+    const { dispatch, action, handleSubmit, children, ...props } = this.props;
+    return (
+      <Formik
+        onSubmit={(...submitArgs) => {
+          const [, { setErrors }] = submitArgs;
+          const dispatchAction = (...actionArgs) =>
+            dispatch({ ...action(...actionArgs), meta: { setErrors } });
```
Comment (Drapegnik) [inline components/common/Form.js]: you override `meta` here

File: components/common/Form.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    action: PropTypes.func.isRequired,
+    handleSubmit: PropTypes.func.isRequired,
+    children: PropTypes.func.isRequired,
+  };
+
+  render() {
+    const { dispatch, action, handleSubmit, children, ...props } = this.props;
+    return (
+      <Formik
+        onSubmit={(...submitArgs) => {
```
Comment (Drapegnik) [inline components/common/Form.js]: clear & simple `onSubmit` example:
```js
    onSubmit={(values, actions) => {
      const submitAction = action(mapValues(values));
      submitAction.meta.formikActions = actions;
      dispatch(submitAction).then(callback);
    }}
```

File: pages/login.js [hunk 13]
```diff
@@ -13,20 +13,15 @@ import request from 'utils/request';
...
 
-const mapDispatchToProps = { signIn: actions.signIn };
-
 class LoginPage extends Component {
   static propTypes = {
     user: PropTypes.shape({}),
     pending: PropTypes.bool.isRequired,
-    errors: PropTypes.shape({}).isRequired,
-    signIn: PropTypes.func.isRequired,
+    // How does url get here?
```
Comment (Drapegnik) [inline pages/login.js]: old `next.js` api: https://github.com/zeit/next.js/blob/master/errors/url-deprecated.md

File: redux/middlewares/form.js [hunk 14]
```diff
@@ -0,0 +1,6 @@
+export default () => next => action => {
+  if (typeof window !== 'undefined' && action.meta && action.meta.setErrors && action.error) {
```
Comment (Drapegnik) [inline redux/middlewares/form.js]: - `typeof window !== 'undefined'` was added only for `ReactGA`, please omit
- lets pass all formik actions to meta:
```js
  if (error && meta && meta.formikActions) {
    const { setErrors } = meta.formikActions;
    setErrors(payload);
  }
```

**Review by Drapegnik — CHANGES_REQUESTED** [review]

File: components/auth/LoginForm.js [hunk 15]
```diff
@@ -35,137 +49,131 @@ const fields = {
...
+        } = url;
+        dispatchAction({ ...userData }, signUp).then(() => Router.pushRoute(next));
+        setSubmitting(false);
+      }}
+    >
+      {({ values, errors, touched, handleSubmit, isSubmitting }) => (
```
Comment (Fusionqq) [inline components/auth/LoginForm.js]: I think u can use 'render' prop instead children

File: components/common/Form.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+          const [, { setErrors }] = submitArgs;
+          const dispatchAction = (...actionArgs) =>
+            dispatch({ ...action(...actionArgs), meta: { setErrors } });
+          handleSubmit(dispatchAction, ...submitArgs);
+        }}
+        {...props}
+      >
+        {children}
```
Comment (Fusionqq) [inline components/common/Form.js]: If u use render prop, u can get rid of children and apply only ```{...props}```

File: components/auth/LoginForm.js [hunk 17]
```diff
@@ -1,29 +1,43 @@
...
+    isSignUpField: true,
     icon: 'user',
-    onlyOnSignUp: true,
   },
   email: {
     icon: 'envelope',
     validator: ({ email }) => required(email) || isEmail(email),
   },
   password: {
+    // ???
```
Comment (Fusionqq) [inline components/auth/LoginForm.js]: 😜

File: components/auth/LoginForm.js (hunk 15 above)
Comment (andemerie) [reply components/auth/LoginForm.js]: Could you argue your proposal?

File: components/auth/FormField.js (hunk 3 above)
Comment (andemerie) [reply components/auth/FormField.js:8]: I'll do it a bit later

**Review by Drapegnik — APPROVED** [review]

**Pull request merged**